File: HackerEarth/March Circuits 2020/Delete and Cut Game/sol.cpp
// #include <iostream>
#include <bits/stdc++.h>

typedef unsigned long long ull;
typedef unsigned long ul;
typedef long long ll;

using namespace std;

ll N, M;

class Node
{
    public:
    ll data;
    ll x = 0; // this holds the component count if I'm part of an edge.
    ll num, low;
    Node* parent = NULL;
    vector<Node*> nodes;

    Node(ll d) : data(d) {}
};

class Graph
{
    public:
    ll counter = 0;
    unordered_map<ll, Node*> nodes;
    vector<pair<ll, ll>> components;

    void addEdge(ll a, ll b)
    {
        if (nodes.find(a) == nodes.end())
        {
            nodes[a] = new Node(a);
        }
        if (nodes.find(b) == nodes.end())
        {
            nodes[b] = new Node(b);
        }
        nodes[a]->nodes.push_back(nodes[b]);
        nodes[b]->nodes.push_back(nodes[a]);
    }

    void bridgesHelper(Node* v, unordered_set<Node*>& visited)
    {
        if (visited.find(v) != visited.end()) return;
        visited.insert(v);
        v->num = counter;
        v->low = counter;
        counter++;
        v->x = 1;

        for(auto& w : v->nodes)
        {
            if (visited.find(w) == visited.end())
            {
                w->parent = v;
                bridgesHelper(w, visited);
                v->x += w->x;
                v->low = min(v->low, w->low);

                if (w->low > v->num)
                {
                    assert(0 < w->x < N);
                    components.push_back({N - w->x, w->x});
                }
            }
            else
            {
                if (v->parent != w)
                {
                    v->low = min(v->low, w->num);
                }
            }
        }
    }

    void findBridges()
    {
        unordered_set<Node*> visited;
        auto nodesIt = nodes.begin();
        bridgesHelper(nodesIt->second, visited);
        assert(nodesIt->second->x == N);
    }

    ll countComponentsHelper(Node* node1, Node* node2, unordered_set<Node*>& visited)
    {
        visited.insert(node1);
        ll rv = 1;
        for(auto w : node1->nodes)
        {
            if (w == node2 || visited.find(w) != visited.end()) continue;
            rv += countComponentsHelper(w, node2, visited);
        }
        return rv;
    }

    pair<ll, ll> countComponents(Node* node1, Node* node2, ll N)
    {
        unordered_set<Node*> visited;
        ll c1 = countComponentsHelper(node1, node2, visited);
        return {c1, N-c1};
    }
};

ll gcdExtended(ll a, ll b, ll *x, ll *y);

// Function to find modulo inverse of b. It returns
// -1 when inverse doesn't
ll modInverse(ll b, ll m)
{
    ll x, y; // used in extended GCD algorithm
    ll g = gcdExtended(b, m, &x, &y);

    // Return -1 if b and m are not co-prime
    if (g != 1)
        return -1;

    // m is added to handle negative x
    return (x%m + m) % m;
}

// Function to compute a/b under modlo m
ll modDivide(ll a, ll b, ll m)
{
    a = a % m;
    ll inv = modInverse(b, m);
    if (inv == -1)
       throw exception();
    else
       return (inv * a) % m;
}

// C function for extended Euclidean Algorithm (used to
// find modular inverse.
ll gcdExtended(ll a, ll b, ll *x, ll *y)
{
    // Base Case
    if (a == 0)
    {
        *x = 0, *y = 1;
        return b;
    }

    ll x1, y1; // To store results of recursive call
    ll gcd = gcdExtended(b%a, a, &x1, &y1);

    // Update x and y using results of recursive
    // call
    *x = y1 - (b/a) * x1;
    *y = x1;

    return gcd;
}

ll m = 1000000007;
int main(int argc, const char** argv)
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    // ll N, M;
    cin >> N >> M;
    Graph graph;
    while (M--)
    {
        ll a, b;
        cin >> a >> b;
        graph.addEdge(a, b);
    }
    graph.findBridges();

    ll aWins = 0, bWins = 0;
    for(auto& c : graph.components)
    {
        assert(c.first + c.second == N);
        if (c.first % 2 == 0 && c.second % 2 == 0) aWins++;
        else bWins++;
    }
    ll t = aWins + bWins;
    if (t == 0) cout << 0 << " " << 0 << endl;
    else cout << modDivide(aWins, t, m) << " " << modDivide(bWins, t, m) << endl;
}

File: HackerEarth/March Circuits 2020/Delete and Cut Game/sol.py
# https://www.hackerearth.com/challenges/competitive/march-circuits-20/algorithm/delete-and-cut-game-91969de1/
import math
import sys
sys.setrecursionlimit(10**5+1)

class Node:
    def __init__(self, data):
        self.data = data
        self.nodes = set()
        self.components = 0 # no. of nodes I'm connected to (in dfs tree)
        self.n = None # discovery time
        self.l = None # lowest node reachable
        self.parent = None

    def __eq__(self, other):
        return self.data == other.data

    def __hash__(self):
        return self.data

    # def __repr__(self):
    #     return f'Node<{self.data, self.parent is None}>'


class Graph:
    def __init__(self):
        self.nodes = {}
        self.counter = 0
        self.bridges = []

    def add_edge(self, a, b):
        if a not in self.nodes:
            self.nodes[a] = Node(a)
        if b not in self.nodes:
            self.nodes[b] = Node(b)
        self.nodes[a].nodes.add(self.nodes[b])
        self.nodes[b].nodes.add(self.nodes[a])

    def bridges_helper(self, v, visited):
        visited.add(v)
        v.n = v.l = self.counter
        self.counter += 1
        # print('c:', v)

        for w in v.nodes:
            if w not in visited:
                w.parent = v
                self.bridges_helper(w, visited)
                v.l = min(w.l, v.l)

                if w.l > v.n:
                    # print('cut edge:', v, w)
                    self.bridges.append((v, w))
            else:
                if v.parent is not None and v.parent != w:
                    # print('updating:', v, w)
                    v.l = min(v.l, w.n)
                continue

    def find_bridges(self):
        visited = set()
        for v in self.nodes.values():
            # print(v)
            if v not in visited:
                self.bridges_helper(v, visited)

    def count_connected_helper(self, node1, node2, visited):
        visited.add(node1)
        rv = 1
        for w in node1.nodes:
            if w == node2 or w in visited:
                continue
            rv += self.count_connected_helper(w, node2, visited)
        return rv

    def count_connected(self, node1, node2):
        visited = set()
        c1 = self.count_connected_helper(node1, node2, visited)
        return c1, N-c1

N, M = map(int, input().split(' '))
graph = Graph()

for _ in range(M):
    a, b = map(int, input().split(' '))
    graph.add_edge(a, b)

# graph.count_connected()
graph.find_bridges()

a_wins = 0
b_wins = 0

for node1, node2 in graph.bridges:
    components1, components2 = graph.count_connected(node1, node2)
    if components1 % 2 == 0 and components2 % 2 == 0:
        a_wins += 1
    else:
        b_wins += 1

def modInverse(b,m):
    g = math.gcd(b, m)
    if (g != 1):
        # print("Inverse doesn't exist")
        return -1
    else:
        # If b and m are relatively prime,
        # then modulo inverse is b^(m-2) mode m
        return pow(b, m - 2, m)

def modDivide(a,b,m):
    a = a % m
    inv = modInverse(b,m)
    if(inv == -1):
        raise Exception
    else:
        return (inv*a) % m

t = a_wins + b_wins
m = 1000000007
if t == 0:
    print(0, 0)
else:
    print(modDivide(a_wins, t, m), modDivide(b_wins, t, m))